### DISCUSSION TITLE
Best Practice for scale of number of subscriptions?

### DISCUSSION THREAD

**Discussion by trevhunter** [ravendb/ravendb#13786] [Q&A] [answered]
In one of our databases, we've decided to use [RavenDB subscriptions](https://ravendb.net/docs/article-page/5.3/csharp/client-api/data-subscriptions/what-are-data-subscriptions) to model async processing that would normally have been implemented using a message queue (simplification: a set of documents represent workflow state and changes to them trigger a subscription for a worker to react to that state change). 

Our implementation of that design has ballooned the number of subscriptions (currently it's now pushing above 40 subscriptions and could grow if we continue that path).

Is scaling the total number of subscriptions a concern? From a performance perspective, or from a management perspective? It certainly makes the "Ongoing Tasks" page in the studio pretty _fun_.

Should we consider refactoring into a smaller or fixed number of subscriptions (i.e. consolidating them and using logic in our workers to branch the processing instead of having such fine grained subscriptions?)

**Comment by ayende** [comment, marked as answer]
There is a (configurable) limit of number of subscriptions connections. 
`Subscriptions.MaxNumberOfConcurrentConnections` - which is set to 1,000 by default (per database / per node).

We have pushed > 100K docs / second via subscriptions, so the ~40 number isn't worrisome from that perspective.
There may be UX issues, admittedly, but no a scaling one. 

Also important is that a subscription that isn't processing document is taking very few resources (basically, just a ping every 3 seconds or so). For subscriptions that do active filtering (checking updated documents but not sending them), there isn't much work done (mostly in memory only).

**Reply by trevhunter** [reply]
Thanks. I should have clarified... I'm not worried about scalability in terms of throughput... we'll handle the volumes of documents fine. The dimension of scaling I was worried about was the number of subscriptions and the implications and limits of that. 

Napkin calculations: 50 subscriptions, 5 worker processes running in concurrent mode (each handling all subscriptions) = 250 concurrent connections or 5 concurrent connections?

**Reply by ayende** [reply]
There is no actual limit for the number of subscriptions you have. The reason we have the connection limit is to be there for _bad_ things. We had an incident where the user opened a subscription _per page view_ on the site. That eventually led to tens of thousands of open subscriptions on the system. We want to detect and block that scenario early, because it doesn't make sense. 

Note that in your model, assuming a 3 node cluster, you'll have the subscriptions spread around, so each node will have ~16 subscriptions and ~84 connections only.

That is very comfortable for us.
